### ISSUE TITLE
Refactor FunDef

### ISSUE THREAD

**Issue by rybern** [stan-dev/stanc3#78]
As I understand it, function definitions will always (and only) appear as the top-level statements in the functions_block list. If this is the case, how about moving FunDef from a general statement to its own type?

```
type 's fun_def = FunDef of
  { fdrt: unsizedtype option
  ; fdname: string
  ; fdargs: fun_arg_decl
  ; fdbody: 's }

type ('e, 's) prog =
  { functions_block: 's fun_def list
  ; ...
```

This has three main advantages:
- It clearly expresses and enforces the invariant that function definitions only appear in the function_block,
- It shrinks the number of variants in statement, so that we don't have to handle the function definition case whenever we match on a statement, which is especially a problem when we don't expect a function definition in e.g. the model block,
- It makes it easier to pull out function definitions from a problem without writing partial functions

I'd be happy to try to work on a pull request if I get the go-ahead.

**Comment by bob-carpenter** [comment]
We want to be thinking forward when they can be defined anywhere using static lexical scoping and/or lambdas with closures.  All the ones that go up front as they do now, just have nothing they can access in scope.  But if we defined such a function in the transformed data block, it should have access to data variables.

At least that's my two cents.

**Comment by VMatthijs** [comment]
The question is whether closures should ever make their way into the MIR though. They'd definitely be in the AST, but we might want to get rid of them through lambda lifting before building the MIR. I'd say we want to keep the MIR as simple as possible. So ideally no closures, higher order functions, structs. Anything that can easily be treated as syntactic sugar should be translated away before the MIR.

**Comment by seantalts** [comment]
So if we got rid of closures, we'd generate a functions block FunDef with more arguments, essentially, and pass in whatever data or transformed data (or stuff that's actually more of a closure like local variables) that it needs, right? Is that acceptable 100% of the time when you might otherwise write a function inline? If so then it seems fine coupled with the appropriate Ast_to_Mir code to do that conversion.

**Comment by rybern** [comment]
Agh, sorry @increasechief! I forgot you mentioned you wanted to try this, I just submitted a pull request. It's possibly incomplete, I'll leave future work to you.

**Comment by increasechief** [comment]
Only in the absence of anyone else caring enough or needing to jump on it. I take no possession here. :)

**Comment by seantalts** [comment]
Hey @VMatthijs, @rybern and I were just talking about this and didn't totally understand the concern you had. After the class would you mind hopping on a call with us for 5-10 minutes to go over it? I tried to model your concern as something like "how do we create the closure to pass to C++" and basically in my head we could do this by e.g. for all closures `f(x)`, generating a `make_f(local1, local2)` function that returns an `f(x)` with the correct stuff closed in.

There are some additional difficulties with MPI, but I don't think keeping FunDefs in the MIR statement helps with that either - we can't just translate into C++ lambdas there. So I'd lean towards taking FunDef out of Mir.statement as the most future-proof move, since we'll be forced to deal with MPI stuff creatively without lambdas anyway.

**Comment by VMatthijs** [comment]
Hey, @seantalts , @rybern , happy to talk about it. I do have some new thoughts about this actually.

**Comment by seantalts** [comment]
Sweet. I'm open the rest of the day. Want to hop in appear.in/quickstan at 2pm? hopefully this catches you both in time :)

**Comment by VMatthijs** [comment]
Yep! Let's do that.

**Comment by seantalts** [comment]
@rybern we talked more - seems like it's a good idea to go ahead with this issue now. Once FunDefs are removed we can stamp the MIR as "approved" for a while longer and me and @bob-carpenter (and whoever else is available) can get back to writing code gen.

**Comment by rybern** [comment]
Okay, sounds good

**Comment by seantalts** [comment]
Closed by #108
